### PR TITLE
Offline users are caching 1.4MB of unused font files

### PR DESCRIPTION
**Pull request by Jodge**
# Description

medic/medic#5260

# Review checklist

- [ ] Readable: Concise, well named, follows the [style guide](https://github.com/medic/medic-docs/blob/master/development/style-guide.md), documented if necessary.
- [ ] Documented: Configuration and user documentation on [medic-docs](https://github.com/medic/medic-docs/)
- [ ] Tested: Unit and/or e2e where appropriate
- [ ] Internationalised: All user facing text
- [ ] Backwards compatible: Works with existing data and configuration or includes a migration. Any breaking changes documented in Changes.md.

# License

The software is provided under AGPL-3.0. Contributions to this project are accepted under the same license.

### CODE REVIEW

**Comment by kennsippell** [comment]
@Jodge Great start! Not sure what our supported browsers are, but I think it's possible that some users may require font files other than woff2. For example, safari and legacy safari users.  

Instead of changing what we actually build, I was thinking we could just change what the service worker caches. So instead of only supporting woff2 font files, it might keep things more open if we only precache woff2 files (via `grunt/service-worker.js`)?  I'm unsure if there are any browsers which support service-workers and don't use woff2, but if there is, add a runtime-caching rule for the other font files, so if they are ever used they will be cached.

**Comment by kennsippell** [comment]
@Jodge Did you Gareth's comment about the italics font from the original issue?

**Comment by Jodge** [comment]
@kennsippell I initially thought of handling this on the `grunt/service-worker.js` but then I had a look at browsers we support i.e. https://github.com/medic/medic-docs/blob/master/installation/supported-software.md and thought why not handle this on `Gruntfile.js` and make it a simple change

For the `font-style: italic` I saw it being used by enketo forms and decided to skip it

**Comment by Jodge** [comment]
> Out of curiosity, why did you pick woff over ttf for the enketo font?

"WOFF fonts often load faster than other formats because they use a compressed version of the structure used by OpenType (OTF) and TrueType (TTF) fonts."

**Comment by kennsippell** [comment]
I didn't test this - but as written, I'd guess that online users still get the `font-awesome.woff2` version of font. Offline users try to download the woff2 version, fail, and then fall back to ttf? Seems we also need a change here which makes ttf the preferred font format for all libraries if we are only going to cache that resource type.

**Comment by garethbowen** [comment]
@kennsippell Looks like it. We'll have to modify font-awesome's path.less so it lists the ttf first, _or_ change the implementation to precache the woff2 version instead.

**Comment by garethbowen** [comment]
Amanda reports that using font sloping is fine so we can do away with both italic fonts, eg:

```css
@font-face {
  font-family: 'Noto';
  src: url(/fonts/NotoSans-Regular.ttf) format("truetype");
  font-weight: normal;
}
@font-face {
  font-family: 'Noto';
  src: url(/fonts/NotoSans-Bold.ttf) format("truetype");
  font-weight: bold;
}
```

File: grunt/service-worker.js [hunk 1]
```diff
@@ -24,8 +24,12 @@ function writeServiceWorkerFile(staticDirectoryPath, rootUrl, outputPath) {
     handleFetch: true,
     staticFileGlobs: [
       path.join(staticDirectoryPath, '{audio,css,img,js,xslt}', '*'),
-      path.join(staticDirectoryPath, 'fonts', '*.ttf'),
       path.join(staticDirectoryPath, 'manifest.json'),
+
+      // Fonts
+      path.join(__dirname, 'fonts', 'fontawesome-webfont.woff2'),
+      path.join(__dirname, 'fonts', 'enketo-icons-v2.woff'),
+      path.join(__dirname, 'fonts', 'Noto*.ttf'),
```
Comment (kennsippell) [inline grunt/service-worker.js]: Won't this cache the italics font? The `less` change below removes our need for those, so let's not precache them.

Comment (Jodge) [reply grunt/service-worker.js]: Yeah that's right, nice catch

File: grunt/service-worker.js [hunk 2]
```diff
@@ -23,8 +23,14 @@ function writeServiceWorkerFile(staticDirectoryPath, rootUrl, outputPath) {
...
     staticFileGlobs: [
-      path.join(staticDirectoryPath, '{audio,css,fonts,img,js,xslt}', '*'),
+      path.join(staticDirectoryPath, '{audio,css,img,js,xslt}', '*'),
       path.join(staticDirectoryPath, 'manifest.json'),
+
+      // Fonts
+      path.join(__dirname, 'fonts', 'fontawesome-webfont.woff2'),
+      path.join(__dirname, 'fonts', 'enketo-icons-v2.woff'),
+      path.join(__dirname, 'fonts', 'NotoSans-Bold.ttf'),
+      path.join(__dirname, 'fonts', 'NotoSans-Regular'),
```
Comment (kennsippell) [inline grunt/service-worker.js]: Need an extension here?

File: webapp/src/css/font.less [hunk 3]
```diff
@@ -6,24 +6,12 @@
...
-  font-family: 'Noto';
-  src: url(/fonts/NotoSans-Italic.ttf) format("truetype");
-  font-weight: normal;
-  font-style: italic;
-}
 @font-face {
   font-family: 'Noto';
   src: url(/fonts/NotoSans-Bold.ttf) format("truetype");
   font-weight: bold;
   font-style: normal;
```
Comment (garethbowen) [inline webapp/src/css/font.less]: You'll need to remove the `font-style` from here and above so Noto is used when the font-style is italic.

File: grunt/service-worker.js (hunk 2 above)
Comment (Jodge) [reply grunt/service-worker.js]: Aargh!,  I am becoming clumsy

**Comment by kennsippell** [comment]
@Jodge  Do you have a quick number of the size of the cache before/after this change? Would be curious if we got the full 1.4MB savings? If you find time.

**Comment by Jodge** [comment]
> @Jodge Do you have a quick number of the reduction in cache size before/after this change? I think as a team, when we do performance changes we should attempt to measure the impact of the changes.

The Cache Storage after this change has reduce from _42.4MB_ to _40.0MB_ so approximately by _2.4MB_

**Pull request merged**